tests/run/critical_sections.pyx

```diff
@@ -0,0 +1,121 @@
+# mode: run
+
+cimport cython
+
+import threading
+
+# This test is most useful on free-threading builds.
+# It should pass on regular builds just by virtue of
+# the GIL.
+
+def test_single_critical_section(n_threads, n_loops):
+    """
+    >>> test_single_critical_section(4, 100)
+    """
+    cdef int count = 0
+    lock = object()
+    barrier = threading.Barrier(n_threads)
+
+    def worker():
+        nonlocal count
+        barrier.wait()
+        for i in range(n_loops):
+            with cython.critical_section(lock):
+                count += i
+
+    threads = [
+        threading.Thread(target=worker)
+        for _ in range(n_threads)
+    ]
+    for t in threads:
+        t.start()
+    for t in threads:
+        t.join()
+
+    expected = ((n_loops * (n_loops - 1))/2)*n_threads
+    assert count == expected, (count, expected)
+
+def test_double_critical_section(n_loops):
+    """
+    >>> test_double_critical_section(100)
+    """
+    cdef int a = 2
+    cdef int b = 3
+    barrier = threading.Barrier(3)
+    finished_event = threading.Event()
+    failed = threading.Event()
+    lock_a = object()
+    lock_b = object()
+
+    def reader1():
+        barrier.wait()
+        while not finished_event.is_set():
+            for _ in range(10):
+                with cython.critical_section(lock_a):
+                    if not (a==2 or a==3):
+                        failed.set()
+
+    def reader2():
+        barrier.wait()
+        while not finished_event.is_set():
+            for _ in range(10):
+                with cython.critical_section(lock_b):
+                    if not (b==2 or b==3):
+                        failed.set()
+
+    def swapper():
+        nonlocal a, b
+        barrier.wait()
+        for _ in range(n_loops):
+            with cython.critical_section(lock_a, lock_b):
+                old_a = a
+                old_b = b
+                a = 1000
+                b = 2000
+                a = old_b
+                b = old_a
```

Comment:
I wonder if this really ends up as code in the binary module. Isn't the C compiler free to remove the intermediate assignments? Maybe add an arbitrary Python operation in between, to be on the safe side?
```suggestion
                # Calling into Python hopefully disrupts C compiler optimisations.
                a = 1000
                b = 2000
                b = 5 + <int> id([])
                a = old_b
                b = old_a
```

Reply:
My worry there is that it's a place that it could break the critical section. (i.e. they aren't atomic if Python decides it needs to use another one in the Python operation)

Reply:
I think I'll rewrite this test to just use `+=` like the others. Since that's something that shouldn't get optimized out.

Reply:
Maybe like this?
```suggestion
                a = 1000
                b = 2000
                # Obfuscate the reset to the C compiler to prevent optimisations.
                if barrier is not None:
                    a = old_b
                    b = old_a
```
But yes, `+= 1` on a pointer deref should not allow optimisations.